src/psi/j/executors/saga.py

```diff
@@ -0,0 +1,222 @@
+        saga_job._jex_uid = job.id
+
+        job_mapping.saga_job = saga_job
+        saga_job.add_callback(rs.STATE, self._state_cb)
+
+        # TODO: the backend attribute is not standard; why do we need it?
+        self._update_job_status(job, JobStatus(JobState.QUEUED, metadata={'backend': self.url}))
+
+        self._update_job_status(job, JobStatus(JobState.ACTIVE))
+
+        saga_job.run()
+
+    def _get_saga_job(self, uid: str) -> rs.job.Job:
+        with self._lock:
+            return self._jobs[uid].saga_job
+
+    def _get_jpsi_job(self, uid: str) -> Job:
+        with self._lock:
+            return self._jobs[uid].jpsi_job
+
+    def list(self) -> List[str]:
+        """
+        Retrieves a list of native IDs corresponding to jobs known to this executor.
+
+        This method does not query the SAGA backend or any of the backends behind SAGA for jobs
+        not submitted through this executor. Furthermore, it simply returns the PSI-J job IDs of
+        the jobs rather than native IDs.
+        Returns
+        -------
+        A list of job IDs.
+        """
+        with self._lock:
+            return list(self._jobs.keys())
+
+    def attach(self, job: Job, native_id: str) -> None:
+        """
+        Attaches a job to a backend job.
+
+        This method does not quite do what it is meant to do at this time.
+
+        Parameters
+        ----------
+        job
+            The job to attach.
+        native_id
+            A PSI-J job ID
+
+        """
+        # This probably needs to be re-worked. I would assume that the meaning of attach()
+        # here is not to attach to a saga job, but to attach to a scheduler job
+
+        # try to find job in known jobs
+        with self._lock:
+            for job_mapping in self._jobs.values():
+                try:
+                    if job_mapping.jpsi_job.status.metadata:
+                        job_native_id = job_mapping.jpsi_job.status.metadata['native_id']
+                        if job_native_id == native_id:
+                            # TODO: what is the meaning here?
+                            # return job_mapping.jpsi_job
+                            return
+                except KeyError:
+                    pass
+
+        # otherwise re-connect from backend:
+        saga_job = None
+        try:
+            saga_job = self._js.get(native_id)
+        except Exception:
+            pass
+
+        if not saga_job:
+            # TODO: There is a discussion on the suitability of cancel() throwing a SubmitException.
+            # TODO: that same discussion would apply here
+            raise SubmitException('Unknown native id: "{}'.format(native_id))
+
+        # we found the job - recreate state
+        state = _STATE_MAP[saga_job.state]
+        if state.final:
+            self._update_job_status(job, JobStatus(state, exit_code=saga_job.exit_code))
+        else:
+            self._update_job_status(job, JobStatus(state))
```

Comment:
before returning, we probably want to update the job mapping in `self._jobs` so that we can update the jpsi_job when the saga_job gets any state notifications?